### DISCUSSION TITLE
How would I do dependency injection?

### DISCUSSION THREAD

**Discussion by JohnBuhanan** [cashapp/better-dynamic-features#126] [General]
Edit: I got a rough scheme working.  Updating code below.  Any suggestions?

How do you guys do dependency injection with this scheme?

Is something like this possible?
```kotlin
// :app
interface CircuitFeature : DynamicApi {
    fun presenterFactories(): Set<Presenter.Factory>
    fun uiFactories(): Set<Ui.Factory>
    fun injectFeature()
}

@Composable
fun ShowSpotify(
    modifier: Modifier = Modifier,
) {
    val installHelper = rememberSplitInstallHelper()

    val isSplitInstalling by installHelper.isInstalling.collectAsState(initial = false)

    val features by remember {
        installHelper.splitInstallManager.dynamicImplementations<CircuitFeature>()
    }.collectAsState(initial = emptyList())

    val circuit = Circuit.Builder()
        .apply {
            features.forEach {
                it.injectFeature()

                for (factory in it.presenterFactories()) {
                    addPresenterFactory(factory)
                }
                for (factory in it.uiFactories()) {
                    addUiFactory(factory)
                }
            }
        }
        .build()

    val backStack = rememberSaveableBackStack(PremiumScreen)
    val navigator = rememberCircuitNavigator(backStack)

    CircuitCompositionLocals(circuit) {
        NavigableCircuitContent(navigator, backStack)
    }
}

// :premium
@DynamicApi
class PremiumFeature : CircuitFeature {
    // Circuit CodeGen with Anvil will "contribute" these factories to DaggerPremiumComponent.
    @Inject
    lateinit var presenterFactories: @JvmSuppressWildcards Set<Presenter.Factory>,

    @Inject
    lateinit var uiFactories: @JvmSuppressWildcards Set<Ui.Factory>,

    override fun presenterFactories(): Set<Presenter.Factory> {
        return presenterFactories
    }

    override fun uiFactories(): Set<Ui.Factory> {
        return uiFactories
    }

    override fun inject() {
        DaggerPremiumComponent
        .factory()
        .create(
// I don't have appComponent coming in yet.
//            (applicationContext as SongifyApplication).appComponent
        )
        .inject(this)
    }
}
```

Neat library guys.  If I end up scaling to something interesting I'd like to come back and show you.
<img width="253" alt="image" src="https://github.com/cashapp/better-dynamic-features/assets/16459434/85d138aa-4637-4ea6-bfa9-3c91ce047a4f">

**Comment by dellisd** [comment]
I think the way that I would model it in a scenario like this is to give each feature module its own component which depends on a shared base component that lives in the base module. The feature module components could adhere to some `@DynamicApi` interface that exposes the actual feature that you're loading.

Instead of injecting each feature when you load it, you could load the feature module components and get a fully-injected instance using that shared `@DynamicApi` component interface. I haven't tried _this_ in practice (our DI graph in Cash is a fair bit more complicated) but I think that's what I would work towards.

This is some really awesome work though and I'm excited to see what you can do with it!

**Reply by JohnBuhanan** [reply]
Could you say more about this?  I am struggling to figure it out :(

Do you mean like this?

```kotlin
interface DFMComponent : DynamicApi {
    val presenterFactories: @JvmSuppressWildcards Set<Presenter.Factory>
    val uiFactories: @JvmSuppressWildcards Set<Ui.Factory>
}

@DynamicImplementation
@SingleIn(FeatureScope::class)
@MergeComponent(
    scope = FeatureScope::class,
)
class HomeComponent : DFMComponent {
    override val presenterFactories: Set<Presenter.Factory>
        get() = TODO("Not yet implemented")
    override val uiFactories: Set<Ui.Factory>
        get() = TODO("Not yet implemented")
}

```

**Comment by JohnBuhanan** [comment]
Here is a video we made showing DFMs downloading:

https://github.com/user-attachments/assets/a48d5f90-796d-4bc1-b224-a5dfbf349a3e

We did all of our work on a branch named `dynamic` in my nephews repo:
https://github.com/JamesBuhanan/Songify/tree/dynamic

I struggle to understand your advice for Components.  I will keep pondering.

Also my app kept crashing when I tried to write `R.drawable.thing`.  Turns out DFMs have to call SplitCompat.install() in order to access resources, so I had to do this ugliness:
<img width="776" alt="image" src="https://github.com/user-attachments/assets/3eaaa70d-cad0-4358-8501-379f87024551">

I don't understand what it is doing, so it is hard for me to tell if this can be made easier for other people.
